### PR TITLE
[WIP] Initial NaiveBayes classifier

### PR DESCRIPTION
**Pull request by ayush1999**
An example usage is:
```
using MLJ
import MLJBase
@load GaussianNBClassifier
using NaiveBayes
using RDatasets
using StatsBase
using Random
# Example 1
iris = dataset("datasets", "iris")

# observations in rows and variables in columns
X = convert(Array, iris[:, 1:4])
n, p = size(X)

y = [species for species in iris[:, 5]]

# how much data use for training
train_frac = 0.9
k = floor(Int, train_frac * n)
idxs = randperm(n)
train = idxs[1:k]
test = idxs[k+1:end];

m = GaussianNBClassifier(classes=unique(y), n=p)

r = MLJBase.fit(m, X[train ,:], y[train], 0)
MLJBase.predict(m, r[1], X[test, :])
```

To Do: 
1. Get this reviewed by @ablaom 
2. Add types to struct
3. Fix a few errors for MultinomialNB

### CODE REVIEW

**Comment by ablaom** [comment]
Many thanks for this work.

Your code needs substantial revisions. I have rewritten the Gaussian part and copied the new code below, after some comments.

## comments

- The model GaussianNBClassifier has only *one* hyperparameter,
  target_type. The parameters `model.c_stats, model.gaussians, model.n_obs` are not hyperparameters
  but parameters to be *learned* by the model being wrapped (analogous
  to weights of a NN). 

- the model constructor GaussianNBClassifier is just a keywork
  constructor for this model. It should have no interaction with the
  package being wrapped.

Regarding fit:

- verbosity argument of fit in wrong position

- I was wrong about NB accepting AbstractVector inputs, so you need to
  collect the adjoint to get a regular matrix.

- need to decode the categorical vector Y as per adding_new_models.md
  instructions. we also need to record *all* levels in the pool of Y,
  not just the ones occurring as elements, and bundle this information
  with the fitresult for use by predict.
  
- not essential, but we can report the values of the learned
  parameters using the report return value of fit
  
Regarding predict:

- need to pad out probabilities for unseen levels in the pool with
  zeros. note that the probabilities reported by NB are not
  normalized; i've reported an issue.
  
## Next steps

- Carefully review code changes and ask questions about points not understood

- Write tests for GaussianNBClassifier along the lines of
  SVMClassifier in src/tests/Scikitlearn.jl in the ScikitLearn branch of
  the MLJModels.jl repo.
  
- Only when these tests pass, refactor MulitinomialClassifier code
  along similar lines. Note this classifier has an extra
  hyperparameter, alpha
  
## new code

````julia
module NaiveBayes_

export GaussianNBClassifier, MultinomialNBClassifier

import MLJBase
using CategoricalArrays
# import ..NaiveBayes
import NaiveBayes

mutable struct GaussianNBClassifier{T} <: MLJBase.Probabilistic{Any} where T
    target_type::Type{T}
end

GaussianNBClassifier(; target_type=Int) = GaussianNBClassifier{target_type}(target_type)

function MLJBase.fit(model::GaussianNBClassifier, verbosity::Int,
                , X
                , Y)

    Xmatrix = MLJBase.matrix(X)' |> collect
    p = size(Xmatrix, 1)

    # find and reorder the levels of y, with observed ones first:
    levels_observed = unique(Y)
    levels_all = levels(Y)
    levels_unobserved = filter(levels_all) do L
        !(L in levels_observed)
    end
    levels_all = vcat(levels_observed, levels_unobserved)

    decoder = MLJBase.CategoricalDecoder(Y)
    y = transform(decoder, Y)

    # NB doesn't accept unobserved levels from the pool of Y
    res = NaiveBayes.GaussianNB(levels_observed, p)
    NaiveBayes.fit(res, Xmatrix, y)

    fitresult = (res, levels_all)
    report = Dict{Symbol,Any}()
    report[:c_counts] = res.c_counts
    report[:c_stats] = res.c_stats
    report[:gaussians] = res.gaussians
    report[:n_obs] = res.n_obs
    
    return fitresult, nothing, report
    
end

function MLJBase.predict(model::GaussianNBClassifier, fitresult, Xnew)

    res, levels_all = fitresult

    Xmatrix = MLJBase.matrix(Xnew)' |> collect
    n = size(Xmatrix, 2)

    levels_observed, logprobs = NaiveBayes.predict_logprobs(res, Xmatrix)

    # NB reports *unnormalized* probabilities!!
    # exponentiate and normalize:
    probs = exp.(logprobs)
    col_sums = sum(probs, dims=1)
    probs = probs ./ col_sums

    # need to give unobserved levels probability zero:
    z = zeros(length(levels_all) - length(levels_observed)) 

    return [UnivariateNominal(levels_all, vcat(probs[:,i], z)) for i in 1:n]
end

#####################################################################
# CODE BELOW IS INCORRECT AND NEEDS REFACTORING ALONG LINES OF CODE #
# ABOVE AFTER THAT CODE PASSES TESTS                                #
#####################################################################

mutable struct MultinomialNBClassifier{T} <: MLJBase.Probabilistic{Any} where T
    target_type::Type{T}
    c_counts
    x_counts
    x_totals
    n_obs::Int
end

function MultinomialNBClassifier(
        ; c_counts
        , x_counts
        , x_totals
        , n_obs)

    m = MultinomialNBClassifier(Int, c_counts, x_counts,
        x_totals, n_obs)

    return m
end

function MultinomialNBClassifier(
    ; classes
    , n)
    m = NaiveBayes.MultinomialNB(classes, n)
    model = MultinomialNBNBClassifier( Int,
        m.c_counts
        ,m.x_counts
        ,m.x_totals
        ,m.n_obs)

    return model
end

function MLJBase.fit(model::MultinomialNBClassifier
                , X
                , Y)

    Xmatrix = MLJBase.matrix(X)'
    res = NaiveBayes.MultinomialNB(model.c_counts, model.x_counts,
    model.x_totals, model.n_obs)

    res = NaiveBayes.fit(res, Float64.(Xmatrix), Y)
    return res, nothing, nothing
end

function MLJBase.predict(model::MultinomialNBClassifier, fitresult, Xnew)
    Xmatrix = MLJBase.matrix(Xnew)'
    return NaiveBayes.predict_logprobs(fitresult, Float64.(Xmatrix)) 
end
    
# metadata:
MLJBase.load_path(::Type{<:GaussianNBClassifier}) = "MLJModels.NaiveBayes_.GaussianNBClassifier" 
MLJBase.package_name(::Type{<:GaussianNBClassifier}) = "GaussianNB"
MLJBase.package_uuid(::Type{<:GaussianNBClassifier}) = "9bbee03b-0db5-5f46-924f-b5c9c21b8c60"
MLJBase.package_url(::Type{<:GaussianNBClassifier}) = "https://github.com/dfdx/NaiveBayes.jl"
MLJBase.is_pure_julia(::Type{<:GaussianNBClassifier}) = :yes
MLJBase.input_kinds(::Type{<:GaussianNBClassifier}) = [:continuous, ]
MLJBase.output_kind(::Type{<:GaussianNBClassifier}) = :multiclass
MLJBase.output_quantity(::Type{<:GaussianNBClassifier}) = :univariate

MLJBase.load_path(::Type{<:MultinomialNBClassifier}) = "MLJModels.NaiveBayes_.MultinomialNBClassifier" 
MLJBase.package_name(::Type{<:MultinomialNBClassifier}) = "MultinomialNB"
MLJBase.package_uuid(::Type{<:MultinomialNBClassifier}) = "9bbee03b-0db5-5f46-924f-b5c9c21b8c60"
MLJBase.package_url(::Type{<:MultinomialNBClassifier}) = "https://github.com/dfdx/NaiveBayes.jl"
MLJBase.is_pure_julia(::Type{<:MultinomialNBClassifier}) = :yes
MLJBase.input_kinds(::Type{<:MultinomialNBClassifier}) = [:multiclass, ]
MLJBase.output_kind(::Type{<:MultinomialNBClassifier}) = :multiclass
MLJBase.output_quantity(::Type{<:MultinomialNBClassifier}) = :univariate

end     #module
```julia

**Comment by ayush1999** [comment]
@ablaom thanks a lot. I have my exams this week, but I'll update the code and add multinomialNB in a couple of days.

**Comment by ayush1999** [comment]
Update: I've opened an issue regarding MultinomialNB in NaiveBayes.jl : https://github.com/dfdx/NaiveBayes.jl/issues/41 .

**Comment by ablaom** [comment]
The iris data set is not an appropriate test set for MultinomialNB because MultinomialNB expects categorical features not continuous ones. Use instead for testing this artificial data set on which your classifier should be very accurate:

````
using CategoricalArrays

generate(n, m) = map(rand(Int, n)) do x mod(x,m) end |> categorical
x1 = generate(1000, 2);
x2 = generate(1000, 3);
x3 = generate(1000, 5);
function f(x)
    if x == 0
        return :zero
    elseif x == 1
        return :one
    else
        return :other
    end
end
X = (x1=x1, x2=x2, x3=x3) # X is tabular
y = map(f, x3)
````

Don't worry about a real data set for now.

**Comment by ayush1999** [comment]
@ablaom Can you please review the changes? (Sorry for the delay, just finished with my exams). For `NaiveBayes.MultinomialNB`, I'm not getting a very good accuracy. Hence I havn't added a test for it as of now.

**Comment by ablaom** [comment]
I have merged after making some changes, and adding tests. Changes include:

- Fixing bug due to NB `predict_logprob` returning target levels in a possibly different order to that passed to fit. 

- dropped `target_type` field for both models, as it is not needed unless we want the `fitresult` type (see MLJ [issue #93](https://github.com/alan-turing-institute/MLJ.jl/issues/93#issue-414875864)). Can put back late if needed

My apologies, the test I recommended for Multinomial was not appropriate. While the inputs for MultinomialBayes are discrete, they should be counts, and so columns of X should be integers, according to our conventions, not categorical vectors. Unfortunately, the tests I added are broken, but I believe this is because of a bug in NaiveBayes.jl. I have raised an [issue](https://github.com/dfdx/NaiveBayes.jl/issues/42#issue-416607368). Let's see what they say. Let's not add the Hybrid case just yet.
